### DISCUSSION TITLE
ESP32-S3 Over The Air (OTA) update with Micropython

### DISCUSSION THREAD

**Discussion by pavelgolikov** [micropython/micropython#10288] [ESP32]
Hi everyone,

I looked for similar discussions, but could not find a clear answer. My apologies if I am asking in the wrong category.

I am looking to set up my esp32-s3 so that it will run micropython and be able to do over the air (OTA) updates. I currently have flashed the latest nightly build of the following micropython version on the device:
https://micropython.org/download/GENERIC_S3/

I noticed that on the same website there is a GENERIC_OTA version, but I do not see GENERIC_S3_OTA version anywhere.

Am I missing something? Is it possible to install GENERIC_OTA on esp32s3 board or should I compile the firmware myself somehow?

Best Regards,

Pavel.

**Comment by git-n-pissed** [comment]
The GENERIC_OTA version will run on the S3, but it will be missing a few features.  At a glance I can see the OTA version isn't built with the USB config enabled (supports the second USB port you would find on ESP32-S3-DevKitC-1 for instance).  HW UART will be disabled and the HW I2C pins will not be defined.  Note that you still have to write the code to make the OTA update happen, unless you use an already available solution like this:

https://github.com/tve/mqboard/blob/master/mqrepl/mqrepl.py

**Reply by pavelgolikov** [reply]
Thank you so much. I might actually need I2C pins for a sensor I want to attach to the device. Do I understand correctly that I can define those I2C pins in software? If so, how would I do that?

Are there any tutorials on how to include/exclude features and how to compile micropython with those features?

**Reply by pavelgolikov** [reply]
I just tried flashing GENERIC_OTA version and I can no longer access the board's REPL with either ampy tool or picocom too. ampy just hangs while picocom shows invalid header message The command I used was:

esptool.py --chip esp32-s3 --port /dev/ttyUSB0 --baud 460800 write_flash -z 0x1000 esp32-ota-20221220-unstable-v1.19.1-782-g699477d12.bin --force

I also tried -z 0 argument for the address of the flash and it still does not work.

I had to use --force flag because otherwise it wouldn't flash it. Am I doing something wrong?

**Reply by git-n-pissed** [reply]
Humm, maybe I lied to you.  I just tried to flash that same image to my S3 and it didn't work.  I think I see now why that is.  In `GENERIC_S3/mpconfigboard.cmake` the IDF target is set to `esp32s3`, and is not set in the GENERIC_OTA build.

`set(IDF_TARGET esp32s3)`

With that said, it does look like you will need to build the firmware to get OTA support for the S3.  Not hard though (I'm doing it for my project).   You need to make sure the `sdkconfig.board` file for your board points to a partition file with the proper settings for your project.  You will note that there are partition files for 16MB and 32MB boards, but no other sizes, so you may need to make a custom one, which also isn't hard.  Some good information in the ESP IDF docs and the MicroPython docs:

https://docs.espressif.com/projects/esp-idf/en/v4.4.3/esp32/api-guides/partition-tables.html
https://docs.espressif.com/projects/esp-idf/en/v4.4.3/esp32/api-reference/system/ota.html
https://docs.micropython.org/en/latest/library/esp32.html#flash-partitions

Also, to clarify, had the GENERIC_OTA build worked, you would have had I2C still, just not HW I2C.  Soft I2C is the default, but it is slower. See the following:

https://docs.micropython.org/en/latest/esp32/quickref.html#software-i2c-bus
https://docs.micropython.org/en/latest/esp32/quickref.html#hardware-i2c-bus
https://docs.espressif.com/projects/esp-idf/en/v4.4.3/esp32s3/api-reference/peripherals/i2c.html
